indexer/src/main/scala/scalding/BaseAlternateNamesJoinIntermediateJob.scala

```diff
@@ -0,0 +1,326 @@
+// Copyright 2014 Foursquare Labs Inc. All Rights Reserved.
+package com.foursquare.twofishes.scalding
+
+import com.twitter.scalding._
+import com.foursquare.twofishes._
+import org.apache.hadoop.io.LongWritable
+import com.twitter.scalding.typed.TypedSink
+import com.foursquare.hadoop.scalding.SpindleSequenceFileSource
+import com.foursquare.twofishes.util.Lists.Implicits._
+import com.foursquare.twofishes.util.NameNormalizer
+
+class BaseAlternateNamesJoinIntermediateJob(
+  name: String,
+  featureSources: Seq[String],
+  altNameSources: Seq[String],
+  args: Args
+) extends TwofishesIntermediateJob(name, args) {
+
+  val features = getJobOutputsAsTypedPipe[LongWritable, GeocodeServingFeature](featureSources).group
+  val altNames = getJobOutputsAsTypedPipe[LongWritable, FeatureNames](altNameSources).group
+
+  val countryInfoCachedFile = getCachedFileByRelativePath("downloaded/countryInfo.txt")
+  @transient lazy val countryNameMap = InMemoryLookupTableHelper.buildCountryNameMap(countryInfoCachedFile)
+  @transient lazy val countryLangMap = InMemoryLookupTableHelper.buildCountryLangsMap(countryInfoCachedFile)
+
+  val customRewritesCachedFile = getCachedFileByRelativePath("custom/rewrites.txt")
+  @transient lazy val nameRewritesMap = InMemoryLookupTableHelper.buildNameRewritesMap(Seq(customRewritesCachedFile))
+
+  val customShortensCachedFile = getCachedFileByRelativePath("custom/shortens.txt")
+  @transient lazy val nameShortensMap = InMemoryLookupTableHelper.buildNameShortensMap(Seq(customShortensCachedFile))
+
+  val customDeletesCachedFile = getCachedFileByRelativePath("custom/deletes.txt")
+  @transient lazy val nameDeletesList = InMemoryLookupTableHelper.buildNameDeletesList(Seq(customDeletesCachedFile))
+
+  def processFeatureName(
+    cc: String,
+    woeType: YahooWoeType,
+    lang: String,
+    name: String,
+    flags: Seq[FeatureNameFlags]
+  ): Seq[FeatureName] = {
+    if (lang != "post") {
+
+      val spaceRe = " +".r
+      def fixName(s: String) = spaceRe.replaceAllIn(s, " ").trim
+
+      // TODO(rahul): actually use flags
+      def doShorten(): Seq[String] = {
+        val shortens = nameShortensMap.getOrElse("*", Nil) ++
+          nameShortensMap.getOrElse(cc, Nil)
+
+        val candidates: Seq[String] = shortens.flatMap(shorten => {
+          val newName = shorten.from.replaceAllIn(name, shorten.to)
+          if (newName != name) {
+            Some(fixName(newName))
+          } else {
+            None
+          }
+        })
+
+        candidates.sortBy(_.length).headOption.toSeq
+      }
+
+      def doRewrites(names: Seq[String]): Seq[String] = {
+        val nameSet = new scala.collection.mutable.HashSet[String]()
+        nameRewritesMap.foreach({case(from, toList) => {
+          names.foreach(name => {
+            toList.foreach(to => {
+              nameSet += from.replaceAllIn(name, to)
+            })
+          })
+        }})
+        nameSet ++= names.map(_.replace("ß", "ss"))
```

Comment:
Yeah. This code (and the code it was copied from) is pretty messy. Didn't want to attempt to cleanup in this iteration.